src/maybe_dangling.rs

```diff
@@ -25,11 +25,133 @@ use ::core::mem::ManuallyDrop as StdMD;
+/// # #[cfg(feature = "nightly-dropck_eyepatch")]
+/// # {
+/// use maybe_dangling::MaybeDangling;
+///
+/// struct Wrapper<'a>(&'a str);
+///
+/// // we pinky-swear not to access the potentially dangling pointer, so `dropck`
+/// // will let us compile this snippet
+/// unsafe impl<#[may_dangle] 'a> Drop for Wrapper<'a> {
+///     fn drop(&mut self) { }
+/// }
+///
+/// fn main() {
+///     let s: String = "I will dangle".into();
+///     let v = MaybeDangling::new(Wrapper(&s));
+///     drop(s); // <- causes the reference in `Wrapper` to dangle
+/// } // <- `v` dropped here
+/// # }
+/// ```
+///
+/// #### Scenario 3: `T` has a destructor and does not use `#[may_dangle]`
+///
+/// This scenario is the same as the previous one, but without `#[maybe_dangle]`.
+/// Drop Check does not know about the internals of `T`'s destructor, so it
+/// can't tell whether the inner reference will be accessed.
+/// Therefore, this will cause a compilation error, even with the
+/// `nightly-dropck_eyepatch` feature enabled.
+///
+/// ```compile_fail
+/// # #[cfg(feature = "nightly-dropck_eyepatch")]
+/// # {
+/// use maybe_dangling::MaybeDangling;
+///
+/// struct Wrapper<'a>(&'a str);
+///
+/// // `dropck` will not know we don't access the dangling pointer and won't let
+/// // us compile this snippet
+/// impl<'a> Drop for Wrapper<'a> {
+///     fn drop(&mut self) { }
+/// }
+///
+/// fn main() {
+///     let s: String = "I will dangle".into();
+///     let v = MaybeDangling::new(Wrapper(&s));
+///     drop(s); // <- causes the reference in `Wrapper` to dangle
+/// } // <- `v` dropped here
+/// # }
+/// # #[cfg(not(feature = "nightly-dropck_eyepatch"))]
+/// # {
+/// #     // to make the snippet not compile when `nightly-dropck_eyepatch`
+/// #     // feature is not enabled
+/// #     fn main() { 0 }
+/// # }
+/// ```
+///
+/// Here a summary of which of the scenarios shown above can be compiled, with
+/// or without the `nightly-dropck_eyepatch` feature enabled:
+///
+/// | `T` | With `nightly-dropck_eyepatch` | Without `nightly-dropck_eyepatch` |
+/// | --- | --- | --- |
+/// | Without destructor | ✅ | ❌ |
+/// | With destructor and with `#[maybe_dangle]` | ✅ | ❌ |
+/// | With destructor and without `#[maybe_dangle]` | ❌ | ❌ |
+///
+/// ### See also
+///
+/// For further information on how automatic drop glue works, see the
+/// "Drop Check" sections of the [`Drop` trait][dropck-std] and the
+/// [Rustonomicon][dropck-nomicon] and the section on how
+/// [drop-checking works with generic parameters][dropck-generics], also found
+/// in the Rustonomicon.
+///
+/// [RFC-1327]: https://rust-lang.github.io/rfcs/1327-dropck-param-eyepatch.html
+/// [dropck-std]: https://doc.rust-lang.org/std/ops/trait.Drop.html#drop-check
+/// [dropck-nomicon]: https://doc.rust-lang.org/nomicon/dropck.html
+/// [dropck-generics]: https://doc.rust-lang.org/nomicon/phantom-data.html#generic-parameters-and-drop-checking
```

Comment:
Let's use permalinks to future-proof against these docs changing in the future and, for instance, the anchors no longer being valid:

```suggestion
/// [dropck-std]: https://doc.rust-lang.org/1.71.0/std/ops/trait.Drop.html#drop-check
/// [dropck-nomicon]: https://doc.rust-lang.org/1.71.0/nomicon/dropck.html
/// [dropck-generics]: https://doc.rust-lang.org/1.71.0/nomicon/phantom-data.html#generic-parameters-and-drop-checking
```